src/mono/mono/metadata/marshal-ilgen.c

```diff
@@ -6193,8 +6267,47 @@ emit_marshal_variant_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 	return conv_arg;
 }
 
+static gboolean
+emit_managed_wrapper_validate_signature (MonoMethodSignature* sig, MonoMarshalSpec** mspecs, MonoError* error)
+{
+	if (mspecs) {
+		for (int i = 0; i < sig->param_count; i ++) {
+			if (mspecs [i + 1] && mspecs [i + 1]->native == MONO_NATIVE_CUSTOM) {
+				if (!mspecs [i + 1]->data.custom_data.custom_name || strlen (mspecs [i + 1]->data.custom_data.custom_name) == 0) {
+					mono_error_set_generic_error (error, "System", "TypeLoadException", "Missing ICustomMarshaler type");
+					return FALSE;
+				}
+
+				switch (sig->params[i]->type) {
+				case MONO_TYPE_OBJECT:
+				case MONO_TYPE_CLASS:
+				case MONO_TYPE_VALUETYPE:
+				case MONO_TYPE_ARRAY:
+				case MONO_TYPE_SZARRAY:
+				case MONO_TYPE_STRING:
+				case MONO_TYPE_BOOLEAN:
+					break;
+				default:
+					mono_error_set_generic_error (error, "System.Runtime.InteropServices", "MarshalDirectiveException", "custom marshalling of type %x is currently not supported", sig->params[i]->type);
+					return FALSE;
+				}
+			} else if (sig->params[i]->type == MONO_TYPE_VALUETYPE) {
+				MonoMarshalType *marshal_type = mono_marshal_load_type_info (mono_class_from_mono_type_internal (sig->params [i]));
+				for (int field_idx = 0; field_idx < marshal_type->num_fields; ++field_idx) {
+					if (marshal_type->fields [field_idx].mspec && marshal_type->fields [field_idx].mspec->native == MONO_NATIVE_CUSTOM) {
+						mono_error_set_generic_error (error, "System", "TypeLoadException", "Value type includes custom marshaled fields");
```

Comment:
```suggestion
						MonoClass *klass = mono_class_from_mono_type_internal (sig->params [i]);
						mono_error_set_type_load_class (error, klass, "Value type includes custom marshaled fields");
```

(could also hoist `mono_class_from_mono_type_internal` from line 6295 above)